### PR TITLE
[TESTING] Improved test coverage slightly (- WIP #53 -)

### LINKED ISSUE #176: Standardize Branch Patterns Across Workflows
**Issue Description:**

There are inconsistencies in branch patterns across various workflow files in the repository. These inconsistencies could lead to workflows not triggering as expected. Standardizing the branch patterns across all workflows will ensure consistent CI/CD behavior.

**References:**

- Original PR: https://github.com/reactive-firewall/multicast/pull/175
- Original Comment: https://github.com/reactive-firewall/multicast/pull/175#discussion_r1823767568

### PR DESCRIPTION
**Pull request by reactive-firewall**
## Related Issues

- [ ] Contributes to #53
- [ ] Contributes to #117
- [x] Closes #176 

---

## Summary by CodeRabbit

This pull request focuses on enhancing the test coverage of the multicast project as part of the ongoing effort to reach comprehensive test coverage outlined in issue [#53](https://github.com/reactive-firewall/multicast/issues/53). It introduces new test modules, updates existing tests, and refines continuous integration workflows to support the expanded test suite. Additionally, it addresses issues [#117](https://github.com/reactive-firewall/multicast/issues/117) and [#176](https://github.com/reactive-firewall/multicast/issues/176) by improving exit code handling and standardizing workflow triggers, respectively.

---

#### `tests/test_exceptions.py`

- **Purpose**: To test the `CommandExecutionError` exception, ensuring proper handling of error messages and exit codes.
- **Key Tests**:
  - Validates that the exception correctly sets the error message and exit code when provided with specific arguments.
  - Confirms that the default exit code is `1` when no exit code is specified.
  - Tests the preservation of the original cause when the exception is raised with a `__cause__`.

#### `tests/test_hear_keyboard_interrupt.py`

- **Purpose**: To verify the application's behavior upon receiving a keyboard interrupt (SIGINT).
- **Key Tests**:
  - Ensures graceful shutdown and appropriate exit code (`130`) when a keyboard interrupt is received.
  - Validates that resources are properly cleaned up after the interrupt.

#### Enhancements in `tests/test_usage.py`

- **Added**: `test_Usage_Error_WHEN_the_help_sub_command_is_called` method.
- **Purpose**: To test that the help output for sub-commands (`HEAR`, `RECV`, `SAY`, `NOOP`) displays correct usage information.
- **Coverage**: Improves test coverage for command-line interface help options.

### 2. Improvements in Test Suite Initialization

- **File**: `tests/__init__.py`
- **Changes**:
  - Imported new test modules (`test_exceptions`, `test_hear_keyboard_interrupt`).
  - Updated `test_cases` to include the new test suites, ensuring they are executed during testing.

#### Addition of Python 3.13 Support

- **Files**:
  - `.github/workflows/Tests.yml`
  - Other workflow files updated accordingly.
- **Purpose**: To ensure compatibility with Python 3.13 and future-proof the project.
- **Changes**:
  - Updated the testing matrix to include Python 3.13.
  - Adjusted jobs to run tests against the new Python version.

#### Explicit Job Permissions

- **Files**:
  - `.github/workflows/Tests.yml`
  - `.github/workflows/Labeler.yml`
  - `.github/workflows/bandit.yml`
  - Others.
- **Purpose**: To enhance security by defining necessary permissions explicitly for each job in the workflows.
- **Changes**:
  - Set permissions for actions such as reading contents, writing statuses, and accessing pull requests.
  - Ensured compliance with the principle of least privilege.

#### Workaround for pip Installation on Windows

- **File**: `.github/workflows/Tests.yml`
- **Purpose**: To address pip installation issues on Windows platforms within the CI environment.
- **Changes**:
  - Added a step to fix pip installation on Windows, improving the reliability of CI tests across different environments.

### 4. Standardization of Workflow Triggers

- **Objective**: Resolves issue [#176](https://github.com/reactive-firewall/multicast/issues/176) by standardizing branch patterns across all GitHub Action workflows.
- **Affected Workflows**:
  - `.github/workflows/Labeler.yml`
  - `.github/workflows/bandit.yml`
  - `.github/workflows/makefile-lint.yml`
  - `.github/workflows/markdown-lint.yml`
  - `.github/workflows/yaml-lint.yml`
- **Changes**:
  - Updated `on.push.branches` and `on.pull_request.branches` to include consistent branch patterns (`"main"`, `"master"`, `"stable"`, `"feature-*"`, `"patch-*"`, `"HOTFIX-*"`) across all workflows.
  - Ensured that workflows are triggered appropriately for all relevant branches.

### 5. Dependabot Configuration Enhancements

- **File**: `.github/dependabot.yml`
- **Changes**:
  - Added several development dependencies to the `allow` list for the `tests/` directory in the `pip` ecosystem.
- **Newly Monitored Dependencies**:
  - `tox`, `virtualenv`, `flake8`, `pep8`, `pytest`, `pytest-checkdocs`, `pytest-cov`, `pytest-enabler`, `pytest-flake8`, `coverage`.
- **Purpose**:
  - To ensure that these testing and development tools are kept up-to-date automatically.
  - Enhances security and stability by proactively managing development dependencies.

### 6. Updates to Coverage Configuration

- **File**: `.coveragerc`
- **Changes**:
  - Added `except ImportError` to the exclusion list in the `[report]` section.
- **Purpose**:
  - To exclude `ImportError` exception handling lines from coverage reports.
  - Focuses coverage metrics on relevant code, improving the accuracy of coverage data.

### 7. Code Clean-up and Minor Enhancements

- **Files Affected**:
  - `multicast/__init__.py`
  - `multicast/__main__.py`
  - `multicast/exceptions.py`
  - `multicast/hear.py`
  - `multicast/recv.py`
  - `multicast/send.py`
- **Changes**:
  - Refactored method signatures to accept `**kwargs`, enhancing flexibility.
  - Improved error handling and resource cleanup.
  - Standardized import statements and module references.
- **Purpose**:
  - To improve code maintainability and readability.
  - To prepare the codebase for future enhancements and refactoring.

---

### Issue [#53](https://github.com/reactive-firewall/multicast/issues/53): Raise Code Coverage to 100%

- **Connection**: This PR significantly contributes to increasing the test coverage of the multicast project.
- **Actions**:
  - Added new test modules covering exceptions and signal handling.
  - Enhanced existing tests to cover edge cases and improve robustness.

### Issue [#117](https://github.com/reactive-firewall/multicast/issues/117): Refactor the Exit Code Handling

- **Connection**: Lays the groundwork for future refactoring by testing current exit code behaviors.
- **Actions**:
  - Validated the handling of exit codes in various scenarios.
  - Ensured that default and specific exit codes behave as expected.

### Issue [#176](https://github.com/reactive-firewall/multicast/issues/176): Standardize Branch Patterns Across Workflows

- **Connection**: This PR closes issue #176 by updating workflow triggers.
- **Actions**:
  - Standardized branch patterns in all GitHub Action workflows.
  - Improved consistency and predictability of CI/CD processes.

### Related Pull Requests

- **PR #118**: Previous work on gathering metrics and improving summaries.
- **PR #148**: Development of the strategic plan for enhancing PR documentation.

---

## Detailed Commit Breakdown

1. **[TESTING] Improved test coverage slightly (- WIP #53 -)**
   - **Changes**:
     - Initial improvements to test coverage.
     - Adjustments to test suite configuration.

2. **[TESTING] Implemented `tests/test_exceptions.py` (- WIP #53 -)**
   - **Changes**:
     - Added tests for `CommandExecutionError`.
     - Improved exception handling coverage.

3. **[TESTING] Implemented `tests/test_hear_keyboard_interrupt.py` (- WIP #53 -)**
   - **Changes**:
     - Added tests for handling keyboard interrupts.
     - Ensured proper cleanup after interruption.

4. **[TESTING] Fix for regression**
   - **Changes**:
     - Addressed regressions introduced by previous changes.
     - Stabilized test executions.

5. **[TESTING] Possible fix for regression Part 2 & Part 3**
   - **Changes**:
     - Continued efforts to resolve test failures.
     - Improved compatibility across different operating systems.

6. **[CONFIG] Fix braindead Windows Python pip issue (- WIP PR #175 -)**
   - **Changes**:
     - Implemented pip installation workaround for Windows.
     - Ensured CI tests pass on Windows environments.

7. **[TESTING] Possible fix for Linux `Ctrl+C` tests (- WIP #53 -)**
   - **Changes**:
     - Fixed issues with signal handling tests on Linux.
     - Ensured consistent behavior across platforms.

8. **[CI] Apply suggestions from code review (- WIP #176 -)**
   - **Changes**:
     - Incorporated feedback to refine workflow configurations.
     - Finalized standardization of workflow triggers.

---

## Impact and Benefits

- **Enhanced Test Coverage**: Moves the project closer to 100% test coverage, improving code reliability and maintainability.
- **Cross-Version Support**: Ensures compatibility with the latest Python release (3.13), future-proofing the project.
- **Improved CI/CD Processes**: Standardized workflows lead to more reliable and secure continuous integration and deployment pipelines.
- **Better Dependency Management**: Proactive monitoring of development dependencies enhances security and stability.
- **Foundation for Future Refactoring**: Validated current behaviors to facilitate upcoming changes, particularly in exit code handling.

---

## Conclusion

This pull request represents a significant step forward in enhancing the robustness and maintainability of the multicast project. By improving test coverage, refining workflows, and laying the groundwork for future enhancements, it contributes to the long-term success and stability of the project.

### CODE REVIEW

File: .github/workflows/Tests.yml [hunk 1]
```diff
@@ -192,6 +221,9 @@ jobs:
       uses: actions/setup-python@v5
       with:
         python-version: ${{ matrix.python-version }}
+    - name: Fix braindead windows ${{ matrix.python-version }} on ${{ matrix.os }}
+      if: ${{ !cancelled() && runner.os == 'Windows' }}
+      run: pip install --upgrade pip
```
Comment (reactive-firewall) [inline .github/workflows/Tests.yml:226]: @coderabbitai, I'm haveing a failures on windows-latest runners in CI related to an out-of-date `pip`.

the logs can be found below for one of the latest attempts:
[Recent CI Run](https://github.com/reactive-firewall/multicast/actions/runs/11603536984/job/32310724820?pr=175)

My observations:
* `pip` is working normally on **non-windows** runners of **same** matrix
  - Thus I suspect this is windows specific.
  - most runners are canceled when the windows-latest runner(s) fails (as configured to do so)
* `pip install --upgrade pip` is not important to the goal of the workflow, rather it is supposed to ensure some version of pip is available
* I am focusing my efforts on the job 'COVERAGE' first, (around line 226 in ".github/workflows/Tests.yml" config)
* also impacts the 'DOCS' job
* this issue has made it clear that the ".github/workflows/Tests.yml" GHA needs to be modularized in a more maintainable way after stabilizing CI

help me debug this failing GHA

Comment (reactive-firewall) [reply .github/workflows/Tests.yml:226]:
```suggestion
    - name: Fix braindead windows ${{ matrix.python-version }} on ${{ matrix.os }}
      if: ${{ !cancelled() && runner.os == 'Windows' }}
      run: python -m pip install --upgrade pip
```
🙈

File: .github/workflows/markdown-lint.yml [hunk 2]
```diff
@@ -1,9 +1,9 @@
 # .github/workflows/markdown-lint.yml
 ---
 name: Markdown Lint
-on:
+on:  # yamllint disable-line rule:truthy
   push:
-    branches: ["main", "master", draft, draft*]
+    branches: ["master", "stable", feature-*, patch-*, HOTFIX-*]
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply .github/workflows/markdown-lint.yml:6]: 🤔 these do differ a bit:
* linters like 'makefile-lint.yml', 'markdown-lint.yml', 'yaml-lint.yml' should be consistent and use `["main", "master", "stable", patch-*, feature-*, HOTFIX-*]`
* `bandit.yml` should be libral on all pull requests but only run on select pushes as with the linters using `["main", "master", "stable", patch-*, feature-*, HOTFIX-*]`
* `codeql-analysis.yml` should be more selective and use only the default branch and stable branches or tags `["main", "master", "stable"]` and `v[0-9].[0-9].[0-9]` respectively
* Tests.yml is rather monolithic right now, and thus must be limited to ration GHA CPU credit usage: only running for the master branch, tags and critical pull requests that target merges to the default branch and "stable"

File: tests/test_hear_keyboard_interrupt.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+__module__ = """tests"""
+
+try:
+	try:
+		import context
+	except Exception as ImportErr:  # pragma: no branch
+		ImportErr = None
+		del ImportErr  # skipcq - cleanup any error leaks early
+		from . import context
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_hear_keyboard_interrupt.py:28]: see #151 and #32 where these should be tracked, please make a detailed note about this suggestion including the simplified code as a comment on #151 to keep track.

File: tests/test_hear_keyboard_interrupt.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+		_fixture_port_num = self._the_test_port
+		try:
+			self.assertIsNotNone(_fixture_port_num)
+			self.assertEqual(type(_fixture_port_num), type(int(0)))
+			_fixture_HEAR_args = [
+				str("{} -m coverage run").format(sys.executable), """multicast""",
+				"""--daemon""", """HEAR""",
+				"""--port""", str(_fixture_port_num),
+				"""--group""", """224.0.0.1"""
+			]
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_hear_keyboard_interrupt.py:59]: 🤔 I agree this should be more readable, however I like the two items per line with the double dash type arguments, so perhaps this would work:
```suggestion
			_fixture_cmd = str("{} -m coverage run").format(sys.executable)
			_fixture_HEAR_args = [
				_fixture_cmd, """multicast""",
				"""--daemon""", """HEAR""",
				"""--port""", str(_fixture_port_num),
				"""--group""", """224.0.0.1"""
			]
```
would this be more readable? why or why not?

File: tests/test_usage.py [hunk 5]
```diff
@@ -651,6 +651,60 @@
...
 
+	def test_Usage_Error_WHEN_the_help_sub_command_is_called(self):
+		"""Test case for multicast* [HEAR|RECV] --help."""
+		theResult = False
+		fail_fixture = str("""multicast [HEAR|RECV] --help == not helpful""")
+		try:
+			inner_fixtures = [
+				("--daemon {}", "HEAR"), ("{}", "HEAR"),
+				("--daemon {}", "RECV"), ("{}", "RECV"), ("{}", "SAY")
+			]
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_usage.py:662]: So I may better consider this suggestion:
Please re-read the file `tests/test_usage.py` focusing on this function (ie `test_Usage_Error_WHEN_the_help_sub_command_is_called`) and then considering your suggestion break-down in detail how the `for` loops would behave before and after the suggested changes. Be thorough.

File: tests/test_usage.py [hunk 6]
```diff
@@ -651,6 +651,60 @@ def test_Usage_Error_WHEN_the_help_command_is_called(self):
...
+						try:
+							if isinstance(theOutputtxt, bytes):
+								theOutputtxt = theOutputtxt.decode('utf8')
+						except UnicodeDecodeError:
+							theOutputtxt = str(repr(bytes(theOutputtxt)))
+						# or simply:
+						self.assertIsNotNone(theOutputtxt)
+						self.assertIn(str("""usage:"""), str(theOutputtxt))
+						if (str("""usage:""") in str(theOutputtxt)):
+							theResult = True or theResult
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_usage.py:694]: good catch, this function should be refactored to set the result to true if the first loop reaches this part, but then use `True and theResult` for subsequent loops so that it only passes if every loop passes and fail if the first loop fails (ie with the `else` clause)

File: .github/workflows/markdown-lint.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply .github/workflows/markdown-lint.yml:6]: open a followup issue for this, and include a detailed summary in the new issue's description.

File: tests/test_exceptions.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+# limitations under the License.
+
+__module__ = """tests"""
+
+try:
+	try:
+		import context
+	except Exception as ImportErr:  # pragma: no branch
+		ImportErr = None
+		del ImportErr  # skipcq - cleanup any error leaks early
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_exceptions.py:27]: see #151 and #32 where these should be tracked, please make a detailed note about this suggestion including the simplified code as a comment on #151 to keep track.

File: tests/test_hear_keyboard_interrupt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_hear_keyboard_interrupt.py:59]: 🤔 I like the f-string improvement but while we're at it, how about this:
```suggestion
			_fixture_cmd = f"{sys.executable} -m coverage run"
			_fixture_HEAR_args = [
				_fixture_cmd, "multicast",
				"--daemon", "HEAR",
				"--port", f"{_fixture_port_num}",
				"--group", "224.0.0.1"
]
```
* basically the same as your idea with similar code length, but the f-string is more inline with the trivial string format goal than `str()` for the integer `_fixture_port_num`
* I typically prefer triple quotes but I admit I over-use them, see related #171

File: tests/test_usage.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_usage.py:694]: what about in the case of an unexpected error?

File: tests/test_usage.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_usage.py:662]: LGTM! please provide a diff with these changes for the suggestion and the related changes needed.

File: .github/workflows/bandit.yml [hunk 8]
```diff
@@ -11,12 +11,12 @@
...
 name: Bandit
-on:
+on:  # yamllint disable-line rule:truthy
   push:
-    branches: [ "master", "stable", feature-*, HOTFIX-* ]
+    branches: ["master", "stable", feature-*, patch-*, HOTFIX-*]
   pull_request:
     # The branches below must be a subset of the branches above
-    branches: [ "master", "stable" ]
+    branches: ["master", "stable"]
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply .github/workflows/bandit.yml:19]: track on #176

File: tests/test_hear_keyboard_interrupt.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_hear_keyboard_interrupt.py:59]: @coderabbitai,
> [!NOTE]
> Clarification: see #171

File: tests/test_usage.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_usage.py:662]: ℹ️ I've added these changes to the next batch of changes.

File: .github/workflows/yaml-lint.yml [hunk 9]
```diff
@@ -3,7 +3,7 @@
 name: YAML Lint
 on:  # yamllint disable-line rule:truthy
   push:
-    branches: ["main", "master", "stable", feature*]
+    branches: ["master", "stable", feature-*, patch-*, HOTFIX-*]
```
Comment (reactive-firewall) [inline .github/workflows/yaml-lint.yml:6]: As part of #176
```suggestion
    branches: ["main", "master", "stable"]
```

File: .github/workflows/yaml-lint.yml [hunk 10]
```diff
@@ -3,7 +3,7 @@
...
   pull_request:
     branches: ["main", "master", "stable"]
```
Comment (reactive-firewall) [inline .github/workflows/yaml-lint.yml:8]: As part of #176
```suggestion
    branches: ["main", "master", "stable", feature-*, patch-*, HOTFIX-*]
```

File: .github/workflows/markdown-lint.yml (hunk 2 above)
Comment (reactive-firewall) [inline .github/workflows/markdown-lint.yml:6]: As part of #176
```suggestion
    branches: ["main", "master", "stable"]
```

File: .github/workflows/markdown-lint.yml [hunk 11]
```diff
@@ -1,9 +1,9 @@
...
   pull_request:
     branches: ["main", "master"]
```
Comment (reactive-firewall) [inline .github/workflows/markdown-lint.yml:8]: As part of #176
```suggestion
    branches: ["main", "master", "stable", feature-*, patch-*, HOTFIX-*]
```

File: .github/workflows/makefile-lint.yml [hunk 12]
```diff
@@ -3,7 +3,7 @@
 name: Makefile Lint
 on:  # yamllint disable-line rule:truthy
   push:
-    branches: ["main", "master", "stable", feature*]
+    branches: ["main", "master", "stable", patch-*, feature-*, HOTFIX-*]
```
Comment (reactive-firewall) [inline .github/workflows/makefile-lint.yml:6]: As part of #176
```suggestion
    branches: ["main", "master", "stable"]
```

File: .github/workflows/makefile-lint.yml [hunk 13]
```diff
@@ -3,7 +3,7 @@
...
   pull_request:
     branches: ["main", "master", "stable"]
```
Comment (reactive-firewall) [inline .github/workflows/makefile-lint.yml:8]: As part of #176
```suggestion
    branches: ["main", "master", "stable", feature-*, patch-*, HOTFIX-*]
```

File: .github/workflows/bandit.yml (hunk 8 above)
Comment (reactive-firewall) [inline .github/workflows/bandit.yml:19]: As part of #176
```suggestion
    branches: ["main", "master", "stable", feature-*, patch-*, HOTFIX-*]
```

File: .github/workflows/bandit.yml [hunk 14]
```diff
@@ -11,12 +11,12 @@
 # https://pypi.org/project/bandit/ is Apache v2.0 licensed, by PyCQA
 
 name: Bandit
-on:
+on:  # yamllint disable-line rule:truthy
   push:
-    branches: [ "master", "stable", feature-*, HOTFIX-* ]
+    branches: ["master", "stable", feature-*, patch-*, HOTFIX-*]
```
Comment (reactive-firewall) [inline .github/workflows/bandit.yml:16]: As part of #176
```suggestion
    branches: ["main", "master", "stable"]
```

**Comment by reactive-firewall** [comment]
@coderabbitai Please recall from our discussion in PR/MR #118, namely [this comment](https://github.com/reactive-firewall/multicast/pull/118#issuecomment-2375448490) and [your response comment](https://github.com/reactive-firewall/multicast/pull/118#issuecomment-2375491898) we figured out how to gather extensive metrics for that PR/MR (#118), and aggregate the information for use in summary improvements.
Please **also** recall that generalizing from the strategy used before in that PR/MR ( #118 ) we developed an improved plan during our discussion in PR/MR #148, namely [this comment](https://github.com/reactive-firewall/multicast/pull/148#issuecomment-2406160141) (but also other related comments)
Similar to [my previous comment](https://github.com/reactive-firewall/multicast/pull/148#issuecomment-2406184695) in PR/MR #148; Please now test out your proposed strategic plan for this PR (i.e. #175), and for the final "Insights Phase" part, use the similar prompt: Generate an in-depth comprehensive overview summarizing the entirety of this PR #175, drawing on the details of relevant commits, issues and related PR/MRs, in a tone appropriate for use as the PR's documentation.

File: multicast/__main__.py [hunk 15]
```diff
@@ -79,52 +80,59 @@
 
 
 try:
-	from . import sys as _sys
-except Exception as impErr:
+	from . import sys
+except ImportError as impErr:
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply multicast/__main__.py:84]: :no_good: This is a special case to ensure all imports are through the `__init__.py` file to ensure proper memory efficiency via the `sys.modules` feature (which was historically done manually by this project before migrating to the python 3 import system). The code here is very specifically intended to import the same `sys` python module object irrespective of client code by forcing single responsibility upon __init__.py for direct imports of shared dependencies. Do you find error with my logic? Please double check the code knowing the intent.

File: tests/test_hear_data_processing.py [hunk 16]
```diff
@@ -74,17 +74,68 @@ def test_multicast_sender_with_no_data(self):
...
+			p = Process(
+				target=multicast.__main__.main,
+				name="HEAR", args=("--daemon", "HEAR", _fixture_HEAR_args,)
+			)
+			p.start()
+			self.assertIsNotNone(p)
+			try:
+				sender = multicast.send.McastSAY()
+				self.assertIsNotNone(sender)
+				sender(group='224.0.0.1', port=_fixture_port_num, ttl=1, data=b'')
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_hear_data_processing.py:119]: The test is checking the corner-case of zero-length byte objects (`b''`) and then following up later with a list of 1 string (e.g. `["STOP"]` which is used to tell the HEAR fixture running on process `p` to perform a server shutdown normally) because lists of strings are the normal (albeit arbitrary) input.

File: multicast/hear.py [hunk 17]
```diff
@@ -518,6 +518,18 @@
...
+		except KeyboardInterrupt as userInterrupt:
+			try:
+				old_sock = server.socket
+				if old_sock:
+					multicast.endSocket(old_sock)
+			finally:
+				raise KeyboardInterrupt("HEAR ended after interruption was signaled.") from userInterrupt
+		finally:
+			server.shutdown()
+		return (_result, None)
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply multicast/hear.py:535]: 🤔 LGTM, with conditions:

* `keyboardInterrupt` is actually tested by sending `signal.SIGINT` (2) so the term "keyboard" can be misleading. Let's use an alternative phrasing.

```suggestion
		server_initialized = False
		server = None
		try:
			with McastServer((HOST, PORT), HearUDPHandler) as server:
				server_initialized = True
				server.serve_forever()
		except KeyboardInterrupt as userInterrupt:
			try:
				if server and server.socket:
					old_sock = server.socket
					multicast.endSocket(old_sock)
			finally:
				raise KeyboardInterrupt(
					f"HEAR has stopped due to interruption signal (was previously listening on {HOST}:{PORT})."
				) from userInterrupt
		finally:
			if server:
				server.shutdown()
		return (server_initialized, None)
```

File: tests/test_fuzz.py [hunk 18]
```diff
@@ -90,26 +90,28 @@
...
 				"""--group""", """'224.0.0.1'""",
 				"""--message""", str("""'{d}'""").format(d=data)
 			]
 			_fixture_HEAR_args = [
+				"""HEAR""",
 				"""--port""", str(_fixture_port_num),
 				"""--groups""", """'224.0.0.1'""",
 				"""--group""", """'224.0.0.1'"""
 			]
 			p = Process(
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_fuzz.py:104]: 🤔 The current implementation is fine. The single quotes are more consistent with how a CLI user would input them and have not caused any issues. Perhaps make a note to explore this further on issue #171 🤷🏻

File: .coderabbit.yaml [hunk 19]
```diff
@@ -19,25 +25,27 @@ reviews:
...
+    - Ensure that that documentation and comments follow
+      [CEP-7](https://gist.github.com/reactive-firewall/123b8a45f1bdeb064079e0524a29ec20)
 
     # Test Code Review Instructions
     - Ensure that test code is automated, comprehensive, and follows testing best practices.
     - Verify that all critical functionality is covered by tests.
-    - Ensure that test code follow
-      [CEP-8](https://gist.github.com/reactive-firewall/d840ee9990e65f302ce2a8d78ebe73f6)
+    - For **test** code, *also* follow
+      [CEP-9](https://gist.github.com/reactive-firewall/d840ee9990e65f302ce2a8d78ebe73f6)
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply .coderabbit.yaml:42]: 👍🏻 Good idea; Please generate a possible implementation to mention the test coverage requirements.

File: tests/test_fuzz.py [hunk 20]
```diff
@@ -207,25 +210,27 @@
...
+					sender(group="224.0.0.1", port=_fixture_port_num, data=["STOP", "Test"])
 					p.join(1)
 				self.assertFalse(p.is_alive())
 			except Exception as _cause:
 				p.join(3)
 				if p.is_alive():
 					p.terminate()
 					p.close()
-				raise unittest.SkipTest(fail_fixture) from _cause
+				raise unittest.SkipTest(sub_fail_fixture) from _cause
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_fuzz.py:227]: brake down your analysis of the suggested refactor for me and refresh my memory about "contextlib"

[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/test_fuzz.py:227]: 🤔 LGTM! With conditions:
* generate appropriate docstring for the function. Be sure to consider project style and conventions.

File: tests/__init__.py [hunk 21]
```diff
@@ -177,6 +185,11 @@ def load_tests(loader, tests, pattern):
 		tests = loader.loadTestsFromTestCase(test_class)
 		suite.addTests(tests)
 	suite.addTests(doctest.DocTestSuite(module=multicast, test_finder=finder))
+	suite.addTests(doctest.DocTestSuite(module=multicast.exceptions, test_finder=finder))
+	suite.addTests(doctest.DocTestSuite(module=multicast.skt, test_finder=finder))
+	suite.addTests(doctest.DocTestSuite(module=multicast.recv, test_finder=finder))
+	suite.addTests(doctest.DocTestSuite(module=multicast.send, test_finder=finder))
+	suite.addTests(doctest.DocTestSuite(module=multicast.hear, test_finder=finder))
```
[the remark this answers is not in the document]
Comment (reactive-firewall) [reply tests/__init__.py:192]: LGTM!

**Comment by reactive-firewall** [comment]
@coderabbitai Please recall from our discussion in PR/MR #118, namely [this comment](https://github.com/reactive-firewall/multicast/pull/118#issuecomment-2375448490) and [your response comment](https://github.com/reactive-firewall/multicast/pull/118#issuecomment-2375491898) we figured out how to gather extensive metrics for that PR/MR (#118), and aggregate the information for use in summary improvements.
Please **also** recall that generalizing from the strategy used before in that PR/MR ( #118 ) we developed an improved plan during our discussion in PR/MR #148, namely [this comment](https://github.com/reactive-firewall/multicast/pull/148#issuecomment-2406160141) (but also other related comments)
Similar to [my previous comment](https://github.com/reactive-firewall/multicast/pull/148#issuecomment-2406184695) in PR/MR #148; Please now test out your proposed strategic plan for this PR (i.e. #175), and for the final "Insights Phase" part, use the similar prompt: Generate an in-depth comprehensive overview summarizing the entirety of this PR #175, drawing on the details of relevant commits, issues and related PR/MRs, in a tone appropriate for use as the PR's documentation.
